nuxeo-core/nuxeo-core-io/src/main/java/org/nuxeo/ecm/core/io/avro/DocumentModelMapper.java

```diff
@@ -76,25 +68,70 @@ public DocumentModel fromAvro(Schema schema, GenericRecord input) {
+            for (Field field : schema.getFields()) {
                 record.put(field.name(), service.toAvro(field.schema(), input));
-            } else {
+            }
+        } else {
+            for (Field field : schema.getFields()) {
                 Property p = input.getProperty(service.decodeName(field.name()));
                 record.put(field.name(), service.toAvro(field.schema(), p));
             }
         }
         return record;
     }
 
+    @SuppressWarnings("unchecked")
+    protected DocumentModel documentModelFromAvro(GenericRecord input) {
+        String path = (String) input.get(Constants.PATH);
+        String type = (String) input.get(Constants.PRIMARY_TYPE);
+        String uuid = (String) input.get(Constants.UUID);
+        String parentId = (String) input.get(Constants.PARENT_ID);
+        String repositoryName = (String) input.get(Constants.REPOSITORY);
+        Boolean isProxy = (Boolean) input.get(Constants.IS_PROXY);
+        DocumentRef parentRef = parentId == null ? null : new IdRef(parentId);
+        Set<String> facets = (Set<String>) input.get(Constants.MIXIN_TYPES);
+        DocumentModelImpl doc = new DocumentModelImpl(null, type, uuid, new Path(path), null, parentRef,
+                null, facets, null, repositoryName, isProxy);
+        doc.setIsVersion((Boolean) input.get(Constants.IS_VERSION));
+        doc.prefetchCurrentLifecycleState((String) input.get(Constants.CURRENT_LIFE_CYCLE_STATE));
+        return doc;
+    }
+
+    protected void documentModelToAvro(Schema schema, DocumentModel doc, GenericRecord record) {
+        record.put(Constants.UUID, doc.getId());
+        record.put(Constants.NAME, doc.getName());
+        record.put(Constants.TITLE, doc.getTitle());
+        record.put(Constants.PATH, doc.getPathAsString());
+        record.put(Constants.REPOSITORY, doc.getRepositoryName());
+        record.put(Constants.PRIMARY_TYPE, doc.getType());
+        DocumentRef parentRef = doc.getParentRef();
+        if (parentRef != null) {
+            record.put(Constants.PARENT_ID, parentRef.toString());
+        }
+        record.put(Constants.CURRENT_LIFE_CYCLE_STATE, doc.getCurrentLifeCycleState());
+        if (doc.isVersion()) {
+            record.put(Constants.VERSION_LABEL, doc.getVersionLabel());
+            record.put(Constants.VERSION_VERSIONABLE_ID, doc.getVersionSeriesId());
+        }
+        record.put(Constants.IS_PROXY, doc.isProxy());
+        record.put(Constants.IS_TRASHED, doc.isTrashed());
+        record.put(Constants.IS_VERSION, doc.isVersion());
+        record.put(Constants.IS_CHECKIN, !doc.isCheckedOut());
+        record.put(Constants.IS_LATEST_VERSION, doc.isLatestVersion());
+        record.put(Constants.IS_LATEST_MAJOR_VERSION, doc.isLatestMajorVersion());
+        record.put(Constants.CHANGE_TOKEN, doc.getChangeToken());
+        if (doc.getPos() != null) {
+            record.put(Constants.IS_PROXY, doc.getPos());
```

Comment:
I don't understand this logic, the fact that a document is a proxy is unrelated to its position.

Reply:
It was an error, Constants.POS should be set actually